platform/machine/oci/cluster.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oci
+
+import (
+	"fmt"
+	"math/rand"
+	"os"
+	"path/filepath"
+
+	"github.com/coreos/pkg/capnslog"
+
+	"github.com/coreos/mantle/platform"
+	"github.com/coreos/mantle/platform/api/oci"
+	"github.com/coreos/mantle/platform/conf"
+)
+
+var (
+	plog = capnslog.NewPackageLogger("github.com/coreos/mantle", "platform/machine/oci")
+)
+
+type cluster struct {
+	*platform.BaseCluster
+	api *oci.API
+}
+
+func NewCluster(opts *oci.Options, rconf *platform.RuntimeConfig) (platform.Cluster, error) {
+	api, err := oci.New(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	bc, err := platform.NewBaseCluster(opts.BaseName, rconf, "")
+	if err != nil {
+		return nil, err
+	}
+
+	oc := &cluster{
+		BaseCluster: bc,
+		api:         api,
+	}
+
+	return oc, nil
+}
+
+func (oc *cluster) vmname() string {
+	b := make([]byte, 5)
+	rand.Read(b)
+	return fmt.Sprintf("%s-%x", oc.Name(), b)
+}
+
+func (oc *cluster) NewMachine(userdata *conf.UserData) (platform.Machine, error) {
+	// the OCI metadata service only exposes the Private IPV4
+	conf, err := oc.RenderUserData(userdata, map[string]string{
+		"$private_ipv4": "${COREOS_OCI_IPV4_PRIVATE_0}",
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// coreos-metadata doesn't produce the private ipv4, adding an override which pulls
+	// it from the vnics portion fo the metadata service
+	conf.AddSystemdUnitDropin("coreos-metadata.service", "oci_override.conf", `[Service]
+Type=oneshot
+Environment=OUTPUT=/run/metadata/coreos
+ExecStart=/usr/bin/mkdir --parent /run/metadata
+ExecStart=/usr/bin/bash -c 'echo "COREOS_OCI_IPV4_PRIVATE_0=$(curl http://169.254.169.254/opc/v1/vnics/ | jq .[].privateIp -r)" > ${OUTPUT}'`, false)
```

Comment:
I'll add the `ExecStart=` line to clobber it because once the information is available on the `vnics` API we can update `coreos-metadata` and drop this dropin.